05-Organizing-and-Executing-Simulations/04-Submitting-Cluster-Jobs.ipynb

```diff
@@ -0,0 +1,611 @@
+{
```

Comment:
<p>I think I am confused here. "Partitioning" is using &gt;1 rank to decrease a single job's simulation time and "Aggregating" is submitting &gt;1 job per cluster to increase state point throughput?</p><p> </p><p>Somehow need to explain that partitioning and aggregating are 2 different strategies for 2 different objectives which may be used independently of each other.</p><p> </p><p>After the second sentence, something like</p><p>"To submit several jobs on a cluster, jobs may be <strong>aggregated</strong> into tuples of a fixed size. To decrease the simulation time of each job, simulations may be <strong>partitioned</strong> across several ranks"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/glotzerlab/hoomd-examples/pull/57/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Reply:
<p>Thanks for sharing your confusion. I need to explain this better. There are actually 3 concepts here, but I oversimplified the explanation.</p><p> </p><ol><li>MPI domain decomposition uses more than one rank to increase the performance of a single simulation in a cluster job.</li><li>MPI partitions use more than one rank to execute multiple simulations in a single cluster job. You can combine partitions with domain decomposition as in this example.</li><li>signac-flow aggregate operations are workflow steps that operate on more than one signac job. For example, you could use this to average properties across many independent simulations in an analysis operation. In this tutorial, I am demonstrating how to use aggregate operations to execute multiple simulations with a single cluster job using MPI partitions.</li></ol><p> </p><p>Does that description help? I could introduce MPI domain decomposition in the first example without partitioning. Then in the second example add partitions used with aggregates. </p><p> </p><p>Would it help to mention that aggregates have other uses or would that be too much information? This tutorial won't cover the details on how to use aggregates except for MPI partitions.</p>

Reply:
<p>That definitely helps, thanks. The point I initially missed is that aggregating is generally anything that operates on more than one job and that partitioning is the particular case where this can be used to submit several simulations using MPI partitions. I think it will be helpful to first explicitly introduce aggregate operation in 1/2 sentences (and maybe link to signac docs for further info) and then say you are going to use an aggregator to use partitions. After that, I think it's clear, I don't think you would need to introduce domain decomposition in the first example.</p>

Reply:
OK, I think I will add domain decomposition to the first example in any case. It makes fewer changes between the two examples and using MPI ranks along with cluster jobs is a typical use-case.